16.0 mig connector search engine

sebastienbeau opened the pull request:
Work in progress
Refactor is comming...

paradoxxxzero committed:
[IMP] connector_search_engine: Use real recompute_json batches with dichotomic retry on fail

sebastienbeau committed:
connector_search_engine: black, isort, prettier

sebastienbeau committed:
connector_search_engine: 16.0 migration

sebastienbeau committed:
connector_search_engine: remove abstract backend

sebastienbeau committed:
connector_search_engine: refactor datamodel

Binding are not anymore an inherit
We have one table with all "binding"
binding can have several state included "to_delete" state

sebastienbeau committed:
connector_search_engine: finish view and improve se.binding design

sebastienbeau committed:
connector_search_engine: remove component dependency

PierrickBrun commented on connector_search_engine/models/se_binding.py:
```diff
@@ -36,25 +27,44 @@ class SeBinding(models.AbstractModel):
         compute="_compute_data_display",
         help="Include this in debug mode to be able to inspect index data.",
     )
-    active = fields.Boolean(default=True)
```
wouldn't it be better to deactivate the binding once it is deleted in the search engine instead of deleting it ?

sebastienbeau replied:
No we can drop it, there is no data to conserve on se.binding, it will avoid to have a growing database

PierrickBrun replied:
agreed, if we want to be able to know if a record was binded this would better be done by having a chatter message on the many2many

sebastienbeau committed:
connector_search_engine: Continue to refactor, clean the module

sebastienbeau committed:
connector_search_engine: Add dedicate menu with specific access right

Now we do not depend anymore of connector

sebastienbeau commented on connector_elasticsearch/models/se_backend.py:
```diff
@@ -0,0 +1,25 @@
+    api_key = fields.Char(help="Elasticsearch Api Key")
+
+    def _get_adapter_list(self):
+        res = super()._get_adapter_list()
```
get_adapter_class

sebastienbeau commented on connector_search_engine/models/se_binding.py:
```diff
@@ -63,138 +85,115 @@ def get_export_data(self):
 
-    # TODO maybe we need to add lock (todo check)
-    def recompute_json(self, force_export=False):
+    def _convert_to_json(self):
```
Replace by a class "get_serializer_class()"

sebastienbeau commented on connector_search_engine/models/se_binding.py:
```diff
@@ -63,138 +85,115 @@ def get_export_data(self):
+                    "flag the binding to be deleted"
+                )
+                continue
+            record.data = record._convert_to_json()
```
```
serializer = record.index_id.get_serializer()
record.data = serializer.serialize(record.record_id)
```

sebastienbeau replied:
validator = record.index_id.get_validator()
record.data = validator.validate(record.record_id)

lmignon commented:
@sebastienbeau Do you plan to work on this in the coming days? Do you need some help to finalize? All this looks very promising.

sebastienbeau committed:
connector_search_engine: clean code for getting search engine adapter

sebastienbeau committed:
connector_search_engine: clean test code, and fix binding deletation

remove the "call_tracking" hack in the context and use mock for test
fix the binding deletation, the id is alway the id of the record, not the id of the binding

sebastienbeau committed:
connector_search_engine: refactor code add concept of serializer and validator

remove dependency on jsonifier, add a default validator

sebastienbeau committed:
connector_search_engine: remove index_bind_ids due to cache issue

sebastienbeau commented:
@lmignon this PR is ready, you can do a review.
I want to make some UI improvement (to make it sexy for end user) but I can do it in a separated PR so we can merge this first step?

lmignon commented:
@sebastienbeau Great! I'll take a look...

lmignon commented on connector_search_engine/tools/adapter.py:
```diff
@@ -2,18 +2,12 @@
-    def match(cls, session, model):
-        # We are a generic exporter; how cool is that?
-        return True  # pragma: no cover
+class SearchEngineAdapter:
```
```suggestion
from abc import ABC, abstractmethod

class SearchEngineAdapter(ABC):
```

lmignon commented on connector_search_engine/tools/adapter.py:
```diff
@@ -2,18 +2,12 @@
+        self.index_record = index_record
 
     def index(self, datas):
         raise NotImplementedError()
```
```suggestion
    @abstractmethod
    def index(self, datas):
        ...
```

Be carefull the three dots are to be preserved...

lmignon committed:
[IMP] connector_search_engine: Mutes logger on queue.job

lmignon committed:
[IMP] connector_search_engine: Add documentation

lmignon committed:
[IMP] connector_search_engine: Improves new implementation

* Uses ABC type to properly define class with abstract methods.
* Improves methods and attributes naming.
* Implements sync mechanism lifecycle.

lmignon committed:
[IMP] connector_search_engine: Add reindex

Add a shortcut method to reindex data

sebastienbeau committed:
Merge pull request #2 from acsone/16.0-mig-connector_search_engine-lmi

16.0 Add documentation and improves naming

sebastienbeau committed:
connector_search_engine: improve UI

- refactor view
- add dashboard for index
- add automatic smart button
- add specific error when converting the record to json

sebastienbeau commented:
Hi all

I have some doubt on the state that we should have on the binding

## Right now here is the list of state

- to_recompute 
=> used to flag a binding to be recomputed (you have done some change on the product for exemple)

- recomputing
=> used to flag a binding when a job have been created for recomputing it
this implied that when a job a created the state on the binding is changed : https://github.com/akretion/search-engine/blob/16.0-mig-connector_search_engine/connector_search_engine/models/se_index.py#L264

- to_export 
=> the compute have been done and the record is ready for the export
       
- exporting
=> used to flag a binding when a job have been created for exporting it
this implied that when a job a created the state on the binding is changed : https://github.com/akretion/search-engine/blob/16.0-mig-connector_search_engine/connector_search_engine/models/se_index.py#L289

- done
=> the item have been exported in the index, everything is ok
              
- invalid_data
=> The validator have raised an error
     
- recompute_error
=> the conversion into json have failed

- to_delete 
=> the item should be deleted in the index (when you delete a record the corresponding binding it set to 'to_delete')
       
- deleting
=> a job have been created for deleting it in the index


In am asking myself if there is not too much state ?

## An alternative can be the following


- to_recompute 
=> used to flag a binding to be recomputed (you have done some change on the product for example), when writing this state a job must be created immediately (so the recomputing state is useless)

- to_export 
=> the compute have been done and the record is ready for the export, when writing this state a job must be created immediately (so the recomputing state is useless)

- done
=> the item have been exported in the index, everything is ok
              
- invalid_data
=> The validator have raised an error
     
- recompute_error
=> the conversion into json have failed

- to_delete 
=> the item should be deleted in the index (when you delete a record the corresponding binding it set to 'to_delete'), when writing this state a job must be created immediately (so the recomputing state is useless)


## pros and cons

option 1:
- when doing a lot of import, reimport, modification just tag the binding instead of creating immediately a lot of job (even if identity key will de-duplicate the job)
- when doing an export have the possibility to do in it batch

options 2 :
- less write on the binding, maybe less concurrency issue ?


Note: we may take our time if we should or not should to this change, we can have a first version with all the detailed state and latter change it if needed

sebastienbeau committed:
connector_search_engine: add action for launching a manual recompute

sebastienbeau committed:
connector_search_engine: improve code, remove useless self.ensure_one

sebastienbeau commented:
After thinking twice and some extra discussion with Laurent.
Let's keep this details state, as it's important to be able to compute by batch and export by batch

sebastienbeau commented:
let's merge it ?

sebastienbeau commented:
@hparfr @lmignon

lmignon commented:
/ocabot merge nobump
